app/services/replacement_parameters_service.rb

```diff
@@ -19,14 +19,15 @@ def process_sensitive_data
   private
 
   def process_replacements_hash(replacements_hash)
-    convert_template_keys_to_string_format_syntax(replacements_hash.keys)
-    body % replacements_hash
-  end
+    # escape existing percent signs
+    body.gsub!(/%(?!{\S*})/, "%%")
 
-  def convert_template_keys_to_string_format_syntax(keys)
-    keys.each{ |key| body.gsub!(/<<\s*#{key}\s*>>/i, "%{#{key}}") }
-  end
+    # replace valid <<key>> with %{key}
+    replacements_hash.each_key { |key| body.gsub!(/<<\s*#{key}\s*>>/i, "%{#{key}}") }
```

Comment:
This makes sense! The computer scientist in me wishes we had a more conceptually holistic way to handle all these cases.

For example, I wonder can we set`.default` on the replacements_hash?

Anyway what you've done here is great. I'm happy to see it merged.

Reply:
Oh! This is something to think about... So this change covers use cases where percentage signs are used that dont look like %{...}. So if there is a `50%` or `your random string is %{45#e`.  What it DOESN'T cover is situations in which there are string in the format of `%{...}` that don't match a replacement parameter. For example `your random string is %{45#e}`. That's still a risk, and I'm not sure how we can mitigate it. Def a downside of doing `body % replacement_params` replacement on a string we don't have control over.